drivers/i2c/i2c_mchp_mss.c

```diff
@@ -0,0 +1,396 @@
+	} while (i2c_status == MSS_I2C_IN_PROGRESS);
+
+	return i2c_status;
+}
+
+static int mss_i2c_read(const struct device *dev, uint8_t serial_addr, uint8_t *read_buffer,
+			uint32_t read_size)
+{
+	struct mss_i2c_data *const data = dev->data;
+	const struct mss_i2c_config *cfg = dev->config;
+
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	data->target_addr = serial_addr << TARGET_ADDR_SHIFT;
+	data->pending_transaction = CONTROLLER_READ_TRANSACTION;
+	data->dir = I2C_MSG_READ;
+	data->controller_rx_buffer = read_buffer;
+	data->controller_rx_size = read_size;
+	data->controller_rx_idx = 0u;
+
+	sys_write8((ctrl | CTRL_STA), cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	return 0;
+}
+
+static int mss_i2c_write(const struct device *dev, uint8_t serial_addr, uint8_t *tx_buffer,
+			 uint32_t tx_num_write)
+{
+	struct mss_i2c_data *const data = dev->data;
+	const struct mss_i2c_config *cfg = dev->config;
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	atomic_t target_status = data->target_status;
+
+	if (data->transaction == NO_TRANSACTION) {
+		data->transaction = CONTROLLER_WRITE_TRANSACTION;
+	}
+
+	data->pending_transaction = CONTROLLER_WRITE_TRANSACTION;
+	data->target_addr = serial_addr << TARGET_ADDR_SHIFT;
+	data->dir = I2C_MSG_WRITE;
+	data->controller_tx_buffer = tx_buffer;
+	data->controller_tx_size = tx_num_write;
+	data->controller_tx_idx = 0u;
+	atomic_set(&data->controller_status, MSS_I2C_IN_PROGRESS);
+
+	if (target_status == MSS_I2C_IN_PROGRESS) {
+		data->is_transaction_pending = CONTROLLER_WRITE_TRANSACTION;
+	} else {
+		sys_write8((ctrl | CTRL_STA), cfg->i2c_base_addr + CORE_I2C_CTRL);
+	}
+
+	if (data->bus_status == MSS_I2C_HOLD_BUS) {
+		sys_write8((ctrl & ~CTRL_SI), cfg->i2c_base_addr + CORE_I2C_CTRL);
+	}
+
+	return 0;
+}
+
+
+static int mss_i2c_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			    uint16_t addr)
+{
+	for (int i = 0; i < num_msgs; i++) {
+		struct i2c_msg *current = &msgs[i];
+
+		if ((current->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+			mss_i2c_read(dev, addr, current->buf, current->len);
+			mss_wait_complete(dev);
+		} else {
+			mss_i2c_write(dev, addr, current->buf, current->len);
+			mss_wait_complete(dev);
+		}
+	}
+
+	return 0;
+}
+
+static const struct i2c_driver_api mss_i2c_driver_api = {
+	.configure = mss_i2c_configure,
+	.transfer = mss_i2c_transfer,
+};
+
+static void mss_i2c_reset(const struct device *dev)
+{
+	const struct mss_i2c_config *cfg = dev->config;
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	sys_write8((ctrl & ~CTRL_ENS1), cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	sys_write8((ctrl | CTRL_ENS1), cfg->i2c_base_addr + CORE_I2C_CTRL);
+}
+
+static int mss_i2c_init(const struct device *dev)
+{
+
+	const struct mss_i2c_config *cfg = dev->config;
+
+	mss_i2c_reset(dev);
+
+	cfg->irq_config_func();
+
+	return 0;
+}
```

Comment:
why `irq_config_func` is needed at all? Can't you just dynamically create the init function that will take care of using `IRQ_INSTALL` / `irq_enable` in there?

Like, in `DEVICE_DT_INST_DEFINE` you instantiate `mss_i2c_init_##inst` and in there you install the IRQs.

Reply:
@carlocaione For this I was following precedent. I found that, for instance, `i2c/i2c_mchp_xec_v2.c`, `i2c/i2c_sam0.c` and many more do it the way I have implemented. I will test what you have suggested today.
Cheers

Reply:
@carlocaione Thanks for the heads up on this one. Tested and working as expected. I think it helps improve the readability of the driver, so thanks again!